File: apps/partner-ios/JYTPartner/PartnerAPI.swift
import Foundation

/// The partner API client — the Swift counterpart of the partner-ui hooks
/// (`apps/partner-ui/src/hooks/api/*`). All requests run against the
/// `/partners/*` surface with the JWT from the Keychain.
actor PartnerAPI {
  static let shared = PartnerAPI()

  private let decoder: JSONDecoder
  private let session: URLSession

  private init() {
    let d = JSONDecoder()
    // ISO 8601 (Medusa's own timestamps) or JS `Date.toString()`. The design
    // routes emit partner_started/finished/completed through
    // `String(new Date(...))` — "Wed Sep 23 2026 09:18:09 GMT+1000
    // (Australian Eastern Standard Time)" — server-locale formatted, so
    // parse it on its own terms (designs/[designId]/route.ts:216).
    d.dateDecodingStrategy = .custom { decoder in
      let container = try decoder.singleValueContainer()
      let raw = try container.decode(String.self)
      for formatter in [ISO8601DateFormatter.fractional, ISO8601DateFormatter.standard] {
        if let date = formatter.date(from: raw) {
          return date
        }
      }
      if let date = Self.parseJavaScriptDate(raw) {
        return date
      }
      throw DecodingError.dataCorruptedError(
        in: container,
        debugDescription: "Unparseable date: \(raw)")
    }
    self.decoder = d
    self.session = URLSession(configuration: .ephemeral)
  }

  /// JS `Date.toString()` — strip the trailing "(Timezone Name)" and parse
  /// "EEE MMM dd yyyy HH:mm:ss 'GMT'Z" with a POSIX locale.
  static let jsDateFormatter: DateFormatter = {
    let f = DateFormatter()
    f.locale = Locale(identifier: "en_US_POSIX")
    f.dateFormat = "EEE MMM dd yyyy HH:mm:ss 'GMT'Z"
    return f
  }()

  static func parseJavaScriptDate(_ raw: String) -> Date? {
    var candidate = raw
    if let paren = raw.range(of: " (") {
      candidate = String(raw[..<paren.lowerBound])
    }
    return jsDateFormatter.date(from: candidate)
  }

  // MARK: - Requests

  /// POST /auth/partner/emailpass — raw login so an unverified email
  /// surfaces as `verificationRequired` instead of an opaque 401.
  struct LoginOutcome {
    let verificationRequired: Bool
    let email: String
  }

  func login(email: String, password: String) async throws -> LoginOutcome {
    let body: [String: String] = ["email": email, "password": password]
    var request = URLRequest(url: Self.makeURL("auth/partner/emailpass"))
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = try JSONSerialization.data(withJSONObject: body)

    let (data, response) = try await session.data(for: request)
    try Self.check(response, data: data)

    let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    if parsed?["verification_required"] as? Bool == true {
      return LoginOutcome(verificationRequired: true, email: email)
    }

    guard let token = parsed?["token"] as? String
            ?? (parsed?["access_token"] as? String) else {
      throw PartnerError.invalidResponse
    }
    Keychain.saveToken(token)
    return LoginOutcome(verificationRequired: false, email: email)
  }

  nonisolated func logout() {
    Keychain.clearToken()
  }

  func me() async throws -> PartnerMe {
    try await get(path: "partners/me")
  }

  func orders(
    kind: String = "design",
    limit: Int = 20,
    offset: Int = 0,
    query: String? = nil
  ) async throws -> PartnerOrderListResponse {
    var path = "partners/orders?kind=\(kind)&limit=\(limit)&offset=\(offset)"
    if let query, !query.isEmpty {
      let encoded = query.addingPercentEncoding(
        withAllowedCharacters: .urlQueryAllowed) ?? query
      path += "&q=\(encoded)"
    }
    return try await get(path: path)
  }

  func order(id: String) async throws -> PartnerOrder {
    struct Wrapper: Codable { let order: PartnerOrder }
    let wrapper: Wrapper = try await get(path: "partners/orders/\(id)")
    return wrapper.order
  }

  // MARK: - Designs & production runs

  func designs(limit: Int = 20, offset: Int = 0, query: String? = nil) async throws -> PartnerDesignListResponse {
    var path = "partners/designs?limit=\(limit)&offset=\(offset)"
    if let query, !query.isEmpty {
      path += "&q=\(query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query)"
    }
    return try await get(path: path)
  }

  func design(id: String) async throws -> DesignDetail {
    struct Wrapper: Codable { let design: DesignDetail }
    let wrapper: Wrapper = try await get(path: "partners/designs/\(id)")
    return wrapper.design
  }

  func productionRuns(designId: String? = nil, limit: Int = 20, offset: Int = 0) async throws -> ProductionRunListResponse {
    var path = "partners/production-runs?limit=\(limit)&offset=\(offset)"
    if let designId {
      path += "&design_id=\(designId)"
    }
    return try await get(path: path)
  }

  func productionRun(id: String) async throws -> ProductionRunDetail {
    try await get(path: "partners/production-runs/\(id)")
  }

  // MARK: - Run lifecycle actions
  // POST /partners/production-runs/:id/{accept,start,finish,complete} —
  // the same mutations the partner-ui milestone hooks wrap
  // (apps/partner-ui/src/hooks/api/partner-production-runs.tsx:128).

  func acceptRun(id: String) async throws {
    try await post(path: "partners/production-runs/\(id)/accept")
  }

  func startRun(id: String) async throws {
    try await post(path: "partners/production-runs/\(id)/start")
  }

  /// `finish` takes an optional notes field (the FinishRunForm's textarea).
  func finishRun(id: String, notes: String?) async throws {
    struct Body: Codable { let notes: String? }
    try await post(
      path: "partners/production-runs/\(id)/finish",
      json: try JSONEncoder().encode(Body(notes: notes))
    )
  }

  /// `complete` output + cost — the CompleteRunForm essentials. The web form
  /// also logs consumptions; those stay on the consumption surfaces, and the
  /// backend accepts a completion without them.
  struct ConsumptionEntry: Codable {
    var inventory_item_id: String?
    var quantity: Double
    var unit_cost: Double?
    var unit_of_measure: String?
    var consumption_type: String?
    var notes: String?
  }

  struct CompleteRunBody: Codable {
    var produced_quantity: Int?
    var rejected_quantity: Int?
    var rejection_reason: String?
    var rejection_notes: String?
    var partner_cost_estimate: Double?
    var cost_type: String?
    var allow_shortfall: Bool?
    var notes: String?
    var consumptions: [ConsumptionEntry]?
  }

  func completeRun(id: String, body: CompleteRunBody) async throws {
    try await post(
      path: "partners/production-runs/\(id)/complete",
      json: try JSONEncoder().encode(body)
    )
  }

  // MARK: - Media upload
  // POST /partners/production-runs/:id/media (multipart `files`) uploads to
  // the storage provider; .../media/attach then merges the URLs into the
  // design's media_files. Same two-step the partner-ui runs.

  struct MediaPart {
    let filename: String
    let mimeType: String
    let data: Data
  }

  func uploadRunMedia(
    runId: String,
    parts: [MediaPart]
  ) async throws -> [UploadedFile] {
    let boundary = "JYTPartnerBoundary-\(UUID().uuidString)"
    var request = URLRequest(url: Self.makeURL("partners/production-runs/\(runId)/media"))
    request.httpMethod = "POST"
    request.setValue(
      "multipart/form-data; boundary=\(boundary)",
      forHTTPHeaderField: "Content-Type"
    )
    if let token = Keychain.loadToken() {
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }
    request.httpBody = Self.multipartBody(boundary: boundary, parts: parts)

    let (data, response) = try await session.data(for: request)
    try Self.check(response, data: data)
    do {
      let decoded = try decoder.decode(UploadFilesResponse.self, from: data)
      return decoded.files
    } catch let error as DecodingError {
      throw PartnerError.decoding(Self.summarize(error))
    } catch {
      throw PartnerError.invalidResponse
    }
  }

  /// Attach freshly uploaded files to the run's design (isThumbnail left
  /// unset — the flagged thumbnail is chosen elsewhere, like the web).
  func attachRunMedia(runId: String, files: [UploadedFile]) async throws {
    let body = AttachMediaBody(
      media_files: files.map { AttachMediaFile(id: $0.id, url: $0.url, isThumbnail: nil) }
    )
    try await post(
      path: "partners/production-runs/\(runId)/media/attach",
      json: try JSONEncoder().encode(body)
    )
  }

  static func multipartBody(boundary: String, parts: [MediaPart]) -> Data {
    var body = Data()
    let line = { (s: String) in body.append(Data(s.utf8)) }
    for part in parts {
      line("--\(boundary)\r\n")
      line("Content-Disposition: form-data; name=\"files\"; filename=\"\(part.filename)\"\r\n")
      line("Content-Type: \(part.mimeType)\r\n\r\n")
      body.append(part.data)
      line("\r\n")
    }
    line("--\(boundary)--\r\n")
    return body
  }

  private func post(path: String, json: Data? = nil) async throws {
    var request = URLRequest(url: Self.makeURL(path))
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let token = Keychain.loadToken() {
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }
    if let json {
      request.httpBody = json
    }
    let (data, response) = try await session.data(for: request)
    try Self.check(response, data: data)
  }

  // MARK: - Plumbing

  /// Build a request URL from "path" or "path?query".
  ///
  /// `URL.appending(path:)` must NOT be used here: it percent-encodes the
  /// `?` of a query string (`orders%3Fkind=design`), turning every
  /// parameterised request into a 404. Path and query are separated
  /// explicitly and set through URLComponents instead.
  static func makeURL(_ path: String) -> URL {
    var components = URLComponents()
    components.scheme = Config.backendURL.scheme
    components.host = Config.backendURL.host
    components.port = Config.backendURL.port

    if let questionMark = path.firstIndex(of: "?") {
      components.path =
        Config.backendURL.path + "/" + String(path[..<questionMark])
      components.query = String(path[path.index(after: questionMark)...])
    } else {
      components.path = Config.backendURL.path + "/" + path
    }
    return components.url ?? Config.backendURL
  }

  private func get<T: Decodable>(path: String) async throws -> T {
    var request = URLRequest(url: Self.makeURL(path))
    request.httpMethod = "GET"
    if let token = Keychain.loadToken() {
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }
    let (data, response) = try await session.data(for: request)
    try Self.check(response, data: data)
    do {
      return try decoder.decode(T.self, from: data)
    } catch let error as DecodingError {
      throw PartnerError.decoding(Self.summarize(error))
    } catch {
      throw PartnerError.invalidResponse
    }
  }

  /// A one-line digest of a decoding failure — keyPath + reason, enough to
  /// name the offending field without dumping the whole context.
  static func summarize(_ error: DecodingError) -> String {
    switch error {
    case .keyNotFound(let key, let context):
      return "missing \(key.stringValue) in \(context.debugDescription.split(separator: " from ").first ?? "")"
    case .typeMismatch(let type, let context):
      return "expected \(type) at \(context.codingPath.map(\.stringValue).joined(separator: "."))"
    case .dataCorrupted(let context):
      let path = context.codingPath.map(\.stringValue).joined(separator: ".")
      return "corrupted value at \(path): \(context.debugDescription.prefix(80))"
    case .valueNotFound(let type, let context):
      return "null \(type) at \(context.codingPath.map(\.stringValue).joined(separator: "."))"
    @unknown default:
      return String(describing: error).prefix(120).description
    }
  }

  private static func check(_ response: URLResponse, data: Data) throws {
    guard let http = response as? HTTPURLResponse else {
      throw PartnerError.invalidResponse
    }
    guard !(200...299).contains(http.statusCode) else { return }
    var message = ""
    if let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
       let msg = parsed["message"] as? String {
      message = msg
    } else if let s = String(data: data, encoding: .utf8), s.count < 300 {
      message = s
    }
    throw PartnerError.http(status: http.statusCode, message: message)
  }
}

private extension ISO8601DateFormatter {
  static let standard: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime]
    return f
  }()

  static let fractional: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return f
  }()
}

File: apps/partner-android/app/src/main/java/com/jyt/partner/api/PartnerApi.kt
package com.jyt.partner.api

import android.content.Context
import android.util.Log
import com.jyt.partner.BuildConfig
import com.jyt.partner.TokenStore
import com.jyt.partner.models.AttachMediaBody
import com.jyt.partner.models.AttachMediaFile
import com.jyt.partner.models.CompleteInventoryOrderBody
import com.jyt.partner.models.DesignDetail
import com.jyt.partner.models.DesignDetailResponse
import com.jyt.partner.models.IncomingDeliveriesResponse
import com.jyt.partner.models.InventoryOrderChargesResponse
import com.jyt.partner.models.PartnerInventoryOrder
import com.jyt.partner.models.PartnerInventoryOrderDetailResponse
import com.jyt.partner.models.PartnerInventoryOrderListResponse
import com.jyt.partner.models.PartnerDesignListResponse
import com.jyt.partner.models.PartnerException
import com.jyt.partner.models.PartnerMe
import com.jyt.partner.models.PartnerOrder
import com.jyt.partner.models.PartnerOrderDetailResponse
import com.jyt.partner.models.PartnerOrderListResponse
import com.jyt.partner.models.ProductionRunDetail
import com.jyt.partner.models.ProductionRunListResponse
import com.jyt.partner.models.ReceiveIncomingBody
import com.jyt.partner.models.UploadFile
import com.jyt.partner.models.UploadFilesResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.serializer
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * The partner API client — the Kotlin counterpart of the iOS app's
 * PartnerAPI actor. All requests run against the /partners surface with
 * the JWT from the encrypted token store.
 *
 * URL building splits path and query explicitly (the iOS app learned this
 * the hard way: appending mangles the `?`); dates stay strings in the
 * models and parse through ApiDate, which takes ISO-8601 AND the JS
 * `Date.toString()` format some design routes emit.
 */
class PartnerApi private constructor(private val context: Context) {

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        coerceInputValues = true
        explicitNulls = false
    }

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    // ── Auth ────────────────────────────────────────────────────────────

    /** POST /auth/partner/emailpass — raw login so an unverified email
     *  surfaces as verificationRequired instead of an opaque 401. */
    data class LoginOutcome(val verificationRequired: Boolean, val email: String)

    @Serializable
    private data class LoginBody(val email: String, val password: String)

    suspend fun login(email: String, password: String): LoginOutcome {
        val map = postObject(
            "auth/partner/emailpass",
            json.encodeToString(LoginBody.serializer(), LoginBody(email, password))
        )
        if (map["verification_required"] as? Boolean == true) {
            return LoginOutcome(verificationRequired = true, email = email)
        }
        val token = map["token"] as? String ?: map["access_token"] as? String
            ?: throw PartnerException.InvalidResponse
        TokenStore.saveToken(context, token)
        return LoginOutcome(verificationRequired = false, email = email)
    }

    fun logout() {
        TokenStore.clearToken(context)
    }

    suspend fun me(): PartnerMe = get("partners/me")

    // ── Orders ──────────────────────────────────────────────────────────

    suspend fun orders(
        kind: String = "design",
        limit: Int = 20,
        offset: Int = 0,
        query: String? = null,
    ): PartnerOrderListResponse {
        var path = "partners/orders?kind=$kind&limit=$limit&offset=$offset"
        if (!query.isNullOrBlank()) path += "&q=${urlEncode(query)}"
        return get(path)
    }

    suspend fun order(id: String): PartnerOrder =
        get<PartnerOrderDetailResponse>("partners/orders/$id").order

    // ── Designs & production runs ───────────────────────────────────────

    suspend fun designs(
        limit: Int = 20,
        offset: Int = 0,
        query: String? = null,
    ): PartnerDesignListResponse {
        var path = "partners/designs?limit=$limit&offset=$offset"
        if (!query.isNullOrBlank()) path += "&q=${urlEncode(query)}"
        return get(path)
    }

    suspend fun design(id: String): DesignDetail =
        get<DesignDetailResponse>("partners/designs/$id").design

    suspend fun productionRuns(
        designId: String? = null,
        limit: Int = 20,
        offset: Int = 0,
    ): ProductionRunListResponse {
        var path = "partners/production-runs?limit=$limit&offset=$offset"
        if (designId != null) path += "&design_id=$designId"
        return get(path)
    }

    suspend fun productionRun(id: String): ProductionRunDetail =
        get("partners/production-runs/$id")

    // ── Run lifecycle actions ────────────────────────────────────────────

    suspend fun acceptRun(id: String) {
        post("partners/production-runs/$id/accept")
    }

    suspend fun startRun(id: String) {
        post("partners/production-runs/$id/start")
    }

    @Serializable
    private data class FinishBody(val notes: String? = null)

    suspend fun finishRun(id: String, notes: String?) {
        post(
            "partners/production-runs/$id/finish",
            json.encodeToString(
                FinishBody.serializer(),
                FinishBody(notes?.takeIf { it.isNotBlank() })
            )
        )
    }

    @Serializable
    data class ConsumptionEntry(
        @SerialName("inventory_item_id") val inventoryItemId: String? = null,
        val quantity: Double,
        @SerialName("unit_cost") val unitCost: Double? = null,
        @SerialName("unit_of_measure") val unitOfMeasure: String? = null,
        @SerialName("consumption_type") val consumptionType: String? = null,
        val notes: String? = null,
    )

    @Serializable
    data class CompleteRunBody(
        @SerialName("produced_quantity") val producedQuantity: Int? = null,
        @SerialName("rejected_quantity") val rejectedQuantity: Int? = null,
        @SerialName("rejection_reason") val rejectionReason: String? = null,
        @SerialName("rejection_notes") val rejectionNotes: String? = null,
        @SerialName("partner_cost_estimate") val partnerCostEstimate: Double? = null,
        @SerialName("cost_type") val costType: String? = null,
        @SerialName("allow_shortfall") val allowShortfall: Boolean? = null,
        val notes: String? = null,
        val consumptions: List<ConsumptionEntry>? = null,
    )

    suspend fun completeRun(id: String, body: CompleteRunBody) {
        post(
            "partners/production-runs/$id/complete",
            json.encodeToString(CompleteRunBody.serializer(), body)
        )
    }

    // ── Media upload ────────────────────────────────────────────────────
    // Two-step like the web: multipart upload → attach to the design.

    class MediaPart(
        val filename: String,
        val mimeType: String,
        val bytes: ByteArray,
    )

    suspend fun uploadRunMedia(runId: String, parts: List<MediaPart>): List<UploadFile> {
        val boundary = "JYTPartnerBoundary-${java.util.UUID.randomUUID()}"
        var payload = ByteArray(0)
        for (part in parts) {
            payload += "--$boundary\r\n".toByteArray()
            payload += "Content-Disposition: form-data; name=\"files\"; filename=\"${part.filename}\"\r\n".toByteArray()
            payload += "Content-Type: ${part.mimeType}\r\n\r\n".toByteArray()
            payload += part.bytes
            payload += "\r\n".toByteArray()
        }
        payload += "--$boundary--\r\n".toByteArray()

        val mediaType = "multipart/form-data; boundary=$boundary".toMediaType()
        val request = baseRequest("partners/production-runs/$runId/media")
            .post(payload.toRequestBody(mediaType)).build()
        val data = execute(request)
        return decodeBody<UploadFilesResponse>(data).files
    }

    /** Attach freshly uploaded files to the run's design. */
    suspend fun attachRunMedia(runId: String, files: List<UploadFile>) {
        val body = AttachMediaBody(
            mediaFiles = files.map { AttachMediaFile(id = it.id, url = it.url) }
        )
        post(
            "partners/production-runs/$runId/media/attach",
            json.encodeToString(AttachMediaBody.serializer(), body)
        )
    }

    // ── Inventory orders ────────────────────────────────────────────────

    suspend fun inventoryOrders(
        limit: Int = 20,
        offset: Int = 0,
        status: String? = null,
        query: String? = null,
    ): PartnerInventoryOrderListResponse {
        var path = "partners/inventory-orders?limit=$limit&offset=$offset"
        if (!status.isNullOrBlank()) path += "&status=${urlEncode(status)}"
        if (!query.isNullOrBlank()) path += "&q=${urlEncode(query)}"
        return get(path)
    }

    suspend fun inventoryOrder(id: String): PartnerInventoryOrder =
        get<PartnerInventoryOrderDetailResponse>("partners/inventory-orders/$id").inventoryOrder

    suspend fun inventoryOrderCharges(id: String): InventoryOrderChargesResponse =
        get("partners/inventory-orders/$id/charges")

    /** Pending → Processing — the partner acknowledges the commission. */
    suspend fun startInventoryOrder(id: String) {
        post("partners/inventory-orders/$id/start")
    }

    /** Processing/Partial → Ready for Delivery — goods packed. */
    suspend fun markInventoryOrderReadyForDelivery(id: String) {
        post("partners/inventory-orders/$id/ready-for-delivery")
    }

    /** The goods receipt: what was actually delivered, per line. */
    suspend fun completeInventoryOrder(id: String, body: CompleteInventoryOrderBody) {
        post(
            "partners/inventory-orders/$id/complete",
            json.encodeToString(CompleteInventoryOrderBody.serializer(), body)
        )
    }

    // ── Incoming deliveries (#2286) ──────────────────────────────────────
    // The other side of the inventory orders: goods delivered TO this
    // partner's warehouse, whoever supplies them.

    /** Default lists only what is still outstanding; all=true includes
     *  fully received orders too. */
    suspend fun incomingDeliveries(all: Boolean = false): IncomingDeliveriesResponse =
        get(if (all) "partners/incoming-deliveries?all=true" else "partners/incoming-deliveries")

    /** The receiving partner states what arrived, per line. */
    suspend fun receiveIncomingDelivery(orderId: String, body: ReceiveIncomingBody) {
        post(
            "partners/incoming-deliveries/$orderId/receive",
            json.encodeToString(ReceiveIncomingBody.serializer(), body)
        )
    }

    // ── Push device tokens ───────────────────────────────────────────────
    // The backend's notification-push provider fans out to whatever the
    // app registers here (POST /partners/device-tokens, which upserts).

    @Serializable
    private data class DeviceTokenBody(
        val token: String,
        val platform: String = "android",
        val app_version: String? = null,
    )

    suspend fun registerDeviceToken(token: String, appVersion: String? = null) {
        post(
            "partners/device-tokens",
            json.encodeToString(
                DeviceTokenBody.serializer(),
                DeviceTokenBody(token, app_version = appVersion)
            )
        )
    }

    @Serializable
    private data class UnregisterBody(val token: String)

    suspend fun unregisterDeviceToken(token: String) {
        val request = baseRequest("partners/device-tokens")
            .delete(
                json.encodeToString(UnregisterBody.serializer(), UnregisterBody(token))
                    .toRequestBody("application/json".toMediaType())
            ).build()
        execute(request)
    }

    // ── Plumbing ────────────────────────────────────────────────────────

    /** Build a request URL from "path" or "path?query" against the
     *  configured backend — path and query are separated explicitly so the
     *  query's `?` is never mangled. */
    private fun url(path: String): String {
        val base = BuildConfig.BACKEND_URL.trimEnd('/')
        return if (path.contains('?')) {
            val parts = path.split('?', limit = 2)
            "$base/${parts[0]}?${parts[1]}"
        } else {
            "$base/$path"
        }
    }

    private fun urlEncode(value: String): String =
        java.net.URLEncoder.encode(value, "UTF-8")

    private fun baseRequest(path: String): Request.Builder {
        val builder = Request.Builder().url(url(path))
        TokenStore.loadToken(context)?.let {
            builder.header("Authorization", "Bearer $it")
        }
        return builder
    }

    private suspend fun execute(request: Request): ByteArray {
        return withContext(Dispatchers.IO) {
            client.newCall(request).execute().use { response ->
                val data = response.body?.bytes() ?: ByteArray(0)
                if (!response.isSuccessful) {
                    throw httpError(response.code, String(data))
                }
                data
            }
        }
    }

    private fun httpError(status: Int, body: String): PartnerException {
        val message = runCatching {
            val obj = Json.parseToJsonElement(body).jsonObject
            (obj["message"] as? JsonPrimitive)?.content
        }.getOrNull() ?: body.take(300)
        return PartnerException.Http(status, message ?: "")
    }

    private suspend inline fun <reified T> decodeBody(data: ByteArray): T {
        return try {
            json.decodeFromString<T>(String(data))
        } catch (e: kotlinx.serialization.SerializationException) {
            Log.w("PartnerApi", "decode failed: ${e.message?.take(160)}")
            throw PartnerException.InvalidResponse
        } catch (e: IllegalArgumentException) {
            throw PartnerException.InvalidResponse
        }
    }

    private suspend inline fun <reified T> get(path: String): T {
        val data = execute(baseRequest(path).get().build())
        return decodeBody<T>(data)
    }

    private suspend fun post(path: String, body: String? = null) {
        val request = baseRequest(path)
            .post((body ?: "{}").toRequestBody("application/json".toMediaType())).build()
        execute(request)
    }

    /** POST returning the parsed top-level JSON object (login). */
    private suspend fun postObject(path: String, body: String): Map<String, Any?> {
        val request = baseRequest(path)
            .post(body.toRequestBody("application/json".toMediaType())).build()
        val data = execute(request)
        val element = runCatching { Json.parseToJsonElement(String(data)) }.getOrNull()
        val obj = element as? JsonObject ?: return emptyMap()
        return obj.mapValues { (_, v) ->
            when (v) {
                is JsonPrimitive -> when {
                    v.isString -> v.content
                    v.booleanOrNull != null -> v.booleanOrNull
                    v.doubleOrNull != null -> v.doubleOrNull
                    else -> null
                }
                else -> null
            }
        }
    }

    companion object {
        @Volatile
        private var instance: PartnerApi? = null

        fun get(context: Context): PartnerApi =
            instance ?: synchronized(this) {
                instance ?: PartnerApi(context.applicationContext).also { instance = it }
            }
    }
}
